application/src/main/java/org/thingsboard/server/service/device/DeviceProvisionServiceImpl.java

```diff
@@ -0,0 +1,326 @@
+            }
+            notify(device, provisionRequest, DataConstants.PROVISION_SUCCESS, true);
+            return new ProvisionResponse(deviceCredentialsService.findDeviceCredentialsByDeviceId(device.getTenantId(), device.getId()), ProvisionResponseStatus.SUCCESS);
+        }, MoreExecutors.directExecutor());
+    }
+
+    private ListenableFuture<ProvisionResponse> createDevice(ProvisionRequest provisionRequest, DeviceProfile profile) {
+        deviceCreationLock.lock();
+        try {
+            return processCreateDevice(provisionRequest, profile);
+        } finally {
+            deviceCreationLock.unlock();
+        }
+    }
+
+    private void notify(Device device, ProvisionRequest provisionRequest, String type, boolean success) {
+        pushProvisionEventToRuleEngine(provisionRequest, device, type);
+        logAction(device.getTenantId(), device.getCustomerId(), device, success, provisionRequest);
+    }
+
+    private ListenableFuture<ProvisionResponse> processCreateDevice(ProvisionRequest provisionRequest, DeviceProfile profile) {
+        Device device = deviceService.findDeviceByTenantIdAndName(profile.getTenantId(), provisionRequest.getDeviceName());
+        if (device == null) {
+            Device savedDevice = saveDevice(provisionRequest, profile);
+
+            deviceStateService.onDeviceAdded(savedDevice);
+            pushDeviceCreatedEventToRuleEngine(savedDevice);
+            notify(savedDevice, provisionRequest, DataConstants.PROVISION_SUCCESS, true);
+
+            return Futures.transform(saveProvisionStateAttribute(savedDevice), input ->
+                    new ProvisionResponse(
+                            getDeviceCredentials(savedDevice),
+                            ProvisionResponseStatus.SUCCESS), MoreExecutors.directExecutor());
+        }
+        log.warn("[{}] The device is already provisioned!", device.getName());
+        notify(device, provisionRequest, DataConstants.PROVISION_FAILURE, false);
+        return Futures.immediateFuture(new ProvisionResponse(null, ProvisionResponseStatus.FAILURE));
+    }
+
+    private ListenableFuture<List<Void>> saveProvisionStateAttribute(Device device) {
+        return attributesService.save(device.getTenantId(), device.getId(), DataConstants.SERVER_SCOPE,
+                Collections.singletonList(new BaseAttributeKvEntry(new StringDataEntry(DEVICE_PROVISION_STATE, PROVISIONED_STATE),
+                        System.currentTimeMillis())));
+    }
+
+    private Device saveDevice(ProvisionRequest provisionRequest, DeviceProfile profile) {
+        Device device = new Device();
+        device.setName(provisionRequest.getDeviceName());
+        device.setType(profile.getName());
+        device.setTenantId(profile.getTenantId());
+        Device savedDevice = deviceService.saveDevice(device);
+        if (!StringUtils.isEmpty(provisionRequest.getCredentialsData().getToken()) ||
```

Comment:
We need to make the main method transactional and pay attention to avoid using DAO layer that is executed in different threads (to avoid locks). Same for TB PE.